numpydoc/xref.py

```diff
@@ -0,0 +1,169 @@
+import re
+from sphinx import addnodes
+
+
+QUALIFIED_NAME_RE = re.compile(
+    # e.g int, numpy.array, ~numpy.array
+    r'^'
+    r'[~\.]?'
+    r'[a-zA-Z_]\w*'
+    r'(?:\.[a-zA-Z_]\w*)*'
+    r'$'
+)
+
+CONTAINER_TYPE_RE = re.compile(
+    # e.g.
+    #  - list[int]
+    #  - dict(str, int)
+    #  - dict[str, int]'
+    #  - tuple(float, float)
+    #  - dict[tuple(str, str), int]'
+    r'^'
+    r'(dict|list|tuple)'
+    r'[\[\(]'
+    r'(.+?(?:,\s*)?)+'
+    r'[\]\)]'
+    r'$'
+)
+
+CONTAINER_SPLIT_RE = re.compile(
+    # splits dict(str, int) into
+    #    ['dict', '[', 'str', ', ', 'int', ']', '']
+    r'(\s*[\[\]\(\),]\s*)'
+)
+
+DOUBLE_QUOTE_SPLIT_RE = re.compile(
+    # splits 'callable ``f(x0, *args)`` or ``f(x0, y0, *args)``' into
+    #    ['callable ', '``f(x0, *args)``', ' or ', '``f(x0, y0, *args)``', '']
+    r'(``.+?``)'
+)
+
+IGNORE = {'of', ' of ', 'either', 'or', 'with', 'in', 'default'}
```

Comment:
Should these settings be wrapped in a class, e.g. `NumpyDocXrefSettings` so that the user can do the following in config:

```python
numpydoc_xref_param_type = NumpyDocXrefSettings()
numpydoc_xref_param_type.ignore.update(['hello'])
```
or perhaps
```python
numpydoc_xref_param_type.add_to_ignore(['hello'])
```

Reply:
It can be configurable, however that set really doesn't affect the final outcome. It is just to minimise useless markup from being inserted.